src/pipewire/pipewire.cc

```diff
@@ -0,0 +1,597 @@
+                             stream_flags, params, aud::n_elems(params)) == 0;
+}
+
+void PipeWireOutput::on_core_event_done(void * data, uint32_t id, int seq)
+{
+    PipeWireOutput * o = static_cast<PipeWireOutput *>(data);
+
+    if (id != PW_ID_CORE || seq != o->m_core_init_seq)
+        return;
+
+    spa_hook_remove(&o->m_registry_listener);
+    spa_hook_remove(&o->m_core_listener);
+
+    o->m_inited = true;
+    pw_thread_loop_signal(o->m_loop, false);
+}
+
+void PipeWireOutput::on_registry_event_global(void * data, uint32_t id, uint32_t permissions,
+                                              const char * type, uint32_t version,
+                                              const struct spa_dict * props)
+{
+    PipeWireOutput * o = static_cast<PipeWireOutput *>(data);
+
+    if (strcmp(type, PW_TYPE_INTERFACE_Node) != 0)
+        return;
+
+    const char * media_class = spa_dict_lookup(props, PW_KEY_MEDIA_CLASS);
+    if (!media_class)
+        return;
+
+    if (strcmp(media_class, "Audio/Sink") != 0)
+        return;
+
+    o->m_has_sinks = true;
+    o->m_core_init_seq = pw_core_sync(o->m_core, PW_ID_CORE, o->m_core_init_seq);
+}
+
+void PipeWireOutput::on_state_changed(void * data, enum pw_stream_state old,
+                                      enum pw_stream_state state, const char * error)
+{
+    PipeWireOutput * o = static_cast<PipeWireOutput *>(data);
+
+    if (o->m_ignore_state_change)
+        return;
+
+    if (state == PW_STREAM_STATE_UNCONNECTED ||
+        state == PW_STREAM_STATE_PAUSED ||
+        state == PW_STREAM_STATE_STREAMING)
+    {
+        pw_thread_loop_signal(o->m_loop, false);
+    }
+}
+
+void PipeWireOutput::on_process(void * data)
+{
+    PipeWireOutput * o = static_cast<PipeWireOutput *>(data);
+    struct pw_buffer * b;
+    struct spa_buffer * buf;
+    void * dst;
+
+    if (!o->m_buffer_at)
+    {
+        pw_thread_loop_signal(o->m_loop, false);
+        return;
+    }
+
+    if (!(b = pw_stream_dequeue_buffer(o->m_stream)))
+    {
+        AUDWARN("PipeWireOutput: out of buffers\n");
+        return;
+    }
+
+    buf = b->buffer;
+
+    if (!(dst = buf->datas[0].data))
+    {
+        AUDWARN("PipeWireOutput: no data pointer\n");
+        return;
+    }
+
+    auto size = aud::min<uint32_t>(buf->datas[0].maxsize, o->m_buffer_at);
+    memcpy(dst, o->m_buffer, size);
+    o->m_buffer_at -= size;
+    memmove(o->m_buffer, o->m_buffer + size, o->m_buffer_at);
+
+    b->buffer->datas[0].chunk->offset = 0;
+    b->buffer->datas[0].chunk->size = o->m_buffer_size;
+    b->buffer->datas[0].chunk->stride = o->m_stride;
+
+    pw_stream_queue_buffer(o->m_stream, b);
+    pw_thread_loop_signal(o->m_loop, false);
+}
+
+void PipeWireOutput::on_drained(void * data)
+{
+    PipeWireOutput * o = static_cast<PipeWireOutput *>(data);
+    pw_thread_loop_signal(o->m_loop, false);
+}
+
+enum spa_audio_format PipeWireOutput::to_pipewire_format(int format)
```

Comment:
Do we need to handle all formats here? The PulseAudio plugin checks less in [to_pulse_format()](https://github.com/audacious-media-player/audacious-plugins/blob/45a0df0ecb479a64d85f982b8cc087b7547e5562/src/pulse/pulse_audio.cc#L369).

Reply:
I don't think we ever send 8-bit or unsigned formats to the output plugins. Supporting them doesn't do any harm though.